[WIP]: Tree-shakable providers documentation

sreevani-ship-it opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[x] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->
No docs existing for new-feature tree-shakable providers.
Issue Number: N/A


## What is the new behavior?
Added tree-shakable provider documentation with a recommendation to use it as the default method.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

mary-poppins commented:
You can preview 7b6b0a7 at https://pr22803-7b6b0a7.ngbuilds.io/.

mary-poppins commented:
You can preview 10ada60 at https://pr22803-10ada60.ngbuilds.io/.

chembu commented:
@jenniferfell , @alxhub : Take a look at the tree-shakable providers doc for V6.

mary-poppins commented:
You can preview 1bd449e at https://pr22803-1bd449e.ngbuilds.io/.

trotyl commented:
Should be able to close #21913 as well

rkirov committed:
fix(compiler): don't typecheck all inputs (#22899)

ngc knows to filter out d.ts inputs, but the logic accidentally
depended on whether it had a previous Program lying around.

Fixing that logic puts ngc on the fast code path, but in that code
path it must be able to merge tsickle EmitResults, so we need to
plumb the tsickle.mergeEmitResults function through all the intervening
APIs.  The bulk of this change is that plumbing.

PR Close #22899

jenniferfell commented on aio/content/guide/styleguide.md:
```diff
@@ -4663,7 +4663,7 @@ Compare with the less preferred `host` metadata alternative.
 
 
-**Do** provide services to the Angular injector at the top-most component where they will be shared.
+**Do** provide services with the app root injector in the `@Injectable` decorator of the service. 
```
Trying to resolve the plural - singular shift. Maybe "Do provide a service with the app root injector in the @Injectable decorator of the service" ?

jenniferfell commented on aio/content/guide/styleguide.md:
```diff
@@ -4685,8 +4685,8 @@ Compare with the less preferred `host` metadata alternative.
-**Why?** When providing the service to a top level component,
-that instance is shared and available to all child components of that top level component.
+**Why?** When providing the service to a root injector,
+that instance is shared and available in every class that needs the service. And provding it in the `@Injectable` decorator allows Angular compiler to perform tree-shaking on services that are not used in an application.
```
Recommend: "...that instance of the service is shared.."

jenniferfell replied:
Last sentence could use some editing. 
- Rewrite to not start sentence with And
- Providing is missing an i
- "It" could be more clear
- "an application"  ->  "the app"
Perhaps "Providing that instance of the service in the @Injectable decorator allows the Angular compiler to perform tree-shaking on services that are not used it the app."

jenniferfell commented on aio/content/guide/providers.md:
```diff
@@ -10,32 +10,40 @@ see the <live-example></live-example>.
-You can provide services to your app by using the `providers` array in an NgModule.
-Consider the default app generated by the CLI. In order to add a user service to it,
-you can generate one by entering the following command in the terminal window:
+Assuming you already have a CLI generated app, create a service using the following CLI command (replace User with the name of your service) in the root project directory:
```
Editorial fine-tuning: 
- "If you have an app generated that was generated with the CLI..."
Or 
- "If you have a CLI-generated app, ..."

jenniferfell replied:
I recommend pulling out the parenthetical. "...create a service using the following CLI command in the root project directory. Replace User with the name of your service."

jenniferfell replied:
What if I don't have a CLI-generated app?

jenniferfell reviewed (changes requested):
Partial review complete

mary-poppins commented:
You can preview c7c0b67 at https://pr22803-c7c0b67.ngbuilds.io/.

mary-poppins commented:
You can preview 95a505f at https://pr22803-95a505f.ngbuilds.io/.

alxhub commented on aio/content/guide/providers.md:
```diff
@@ -10,32 +10,62 @@ see the <live-example></live-example>.
+
+<code-example>
+
+@Injectable({
```
Example: user.service.0.ts

alxhub commented on aio/content/guide/providers.md:
```diff
@@ -10,32 +10,62 @@ see the <live-example></live-example>.
+It's also possible to specify that a service should be provided in a particular `@NgModule`. For example, if you don't want the `UserService` to be available to applications unless they import a `UserModule` you've created, you can specify that the service should be provided in the module:
 
+<code-example>
+@Injectable({ 
```
Example: user.service.1.ts

alxhub commented on aio/content/guide/providers.md:
```diff
@@ -10,32 +10,62 @@ see the <live-example></live-example>.
+
+<code-example>
+@NgModule({ 
+    
```
Example: user.module.ts

alxhub commented on aio/content/guide/styleguide.md:
```diff
@@ -4715,17 +4715,18 @@ that instance is shared and available to all child components of that top level
-<code-tabs>
-
-  <code-pane title="app/app.component.ts" path="styleguide/src/07-03/app/app.component.ts">
+<code-example>
```
Example: hero.service.ts

sreevani-ship-it committed:
docs: update contents with code examples

sreevani-ship-it committed:
docs: add tree-shakable providers code examples to doc

sreevani-ship-it committed:
docs: fix conflicts

sreevani-ship-it committed:
docs: fix conflicts

alfaproject commented on aio/content/guide/dependency-injection-in-action.md:
```diff
@@ -11,7 +11,11 @@ of the code in this cookbook.
 ## Application-wide dependencies
-Register providers for dependencies used throughout the application in the root application component, `AppComponent`.
+
+From V6 onwards, providers for application-wide dependencues should be defined by specifying `providedIn: 'root'` on their `@Injectable()` decorator (in the case of services) or at `InjectionToken` construction (in the case of tokens being provided).
```
dependencies*

chembu commented:
New PR: https://github.com/angular/angular/pull/23027

jenniferfell commented:
@alfaproject Thanks. Good catch!

jenniferfell commented:
Moved work to https://github.com/angular/angular/pull/23027
